services/source/updater.c

```diff
@@ -0,0 +1,355 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+/**
+ * @file updater.c
+ * @author Robert Taylor
+ * @date 2021-07-20
+ */
+
+#include "updater.h"
+#include "bl_eeprom.h"
+#include "bl_flash.h"
+#include "services.h"
+#include <FreeRTOS.h>
+#include <csp/csp.h>
+#include <csp/csp_endian.h>
+#include <main/system.h>
+#include <os_task.h>
+#include "service_utilities.h"
+#include <FreeRTOS-Plus-CLI/FreeRTOS_CLI.h>
+#include "printf.h"
+#include "privileged_functions.h"
+
+static struct __attribute__((packed)) {
+    uint32_t start_address;
+    uint32_t size;
+    uint32_t next_address;
+    bool initialized;
+} update = {0};
+
+static char *error = "None";
+/**
+ * hex2int
+ * take a hex string and convert it to a 32bit number (max 8 hex digits)
+ */
+uint32_t hex2int(const char *hex) {
+    uint32_t val = 0;
+    while (*hex) {
+        // get current character then increment
+        uint8_t byte = *hex++;
+        // transform hex character to the 4bit equivalent number, using the ascii table indexes
+        if (byte >= '0' && byte <= '9') byte = byte - '0';
+        else if (byte >= 'a' && byte <='f') byte = byte - 'a' + 10;
+        else if (byte >= 'A' && byte <='F') byte = byte - 'A' + 10;
+        // shift 4 to make space for new digit, and add the 4 bits of the new digit
+        val = (val << 4) | (byte & 0xF);
+    }
+    return val;
+}
+
+static BaseType_t prvAddrCommand(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString) {
+    int32_t image_param_len = 0;
+    const char *image = FreeRTOS_CLIGetParameter(
+                    /* The command string itself. */
+                    pcCommandString,
+                    /* Return the next parameter. */
+                    1,
+                    /* Store the parameter string length. */
+                    &image_param_len);
+    if (image_param_len > 1) {
+        snprintf(pcWriteBuffer, xWriteBufferLen, "Invalid Image Type\n");
+        return pdFALSE;
+    }
+
+    int32_t addr_str_len;
+    const char *addr_str = FreeRTOS_CLIGetParameter(
+                    /* The command string itself. */
+                    pcCommandString,
+                    /* Return the next parameter. */
+                    2,
+                    /* Store the parameter string length. */
+                    &addr_str_len);
+
+    if (addr_str_len > 10) {
+        snprintf(pcWriteBuffer, xWriteBufferLen, "Hex string too long\n");
+        return pdFALSE;
+    }
+
+    const char *hex_values = addr_str + 2;
```

Comment:
I assume this is for "0x" - do you know for sure it's there?

Reply:
Since it's just a debug function I didn't pay too much attention to error handling. However, hex2int does do this with error checking now, so I'll remove it